deps/oculus/src/ovrsession.cpp

```diff
@@ -0,0 +1,436 @@
+    }
+  }
+}
+
+NAN_METHOD(OVRSession::Submit)
+{
+
+  if (info.Length() != 5)
+  {
+    Nan::ThrowError("Wrong number of arguments.");
+    return;
+  }
+
+  if (!(info[0]->IsObject() && info[1]->IsNumber()))
+  {
+    Nan::ThrowError("Expected arguments (object, number).");
+    return;
+  }
+
+  WebGLRenderingContext *gl = node::ObjectWrap::Unwrap<WebGLRenderingContext>(Local<Object>::Cast(info[0]));
+  GLuint textureSource = info[1]->Uint32Value();
+  GLuint fboIdSource = info[2]->Uint32Value();
+  int canvasWidth = info[3]->Uint32Value();
+  int canvasHeight = info[4]->Uint32Value();
+
+  ovrPosef *eyeRenderPoses = &*ObjectWrap::Unwrap<OVRSession>(info.Holder())->eyeRenderPoses;
+  int *frameIndex = &ObjectWrap::Unwrap<OVRSession>(info.Holder())->frameIndex;
+  GLuint fboId = ObjectWrap::Unwrap<OVRSession>(info.Holder())->fboId;
+  ovrSession session = *ObjectWrap::Unwrap<OVRSession>(info.Holder())->self_;
+  double sensorSampleTime = ObjectWrap::Unwrap<OVRSession>(info.Holder())->sensorSampleTime;
+  EyeSwapChain *eyes = ObjectWrap::Unwrap<OVRSession>(info.Holder())->eyes;
+  ovrHmdDesc hmdDesc = ObjectWrap::Unwrap<OVRSession>(info.Holder())->hmdDesc;
+
+  // Eye textures clear.
+  GLuint colorTextureId;
+  GLuint depthTextureId;
+  ovrTimewarpProjectionDesc posTimewarpProjectionDesc = {};
+  glClearColor(1.0, 0.0, 1.0, 1.0);
+
+  for (int eye = 0; eye < 2; ++eye) {
+
+    {
+      int curIndex;
+      ovr_GetTextureSwapChainCurrentIndex(session, eyes[eye].ColorTextureChain, &curIndex);
+      ovr_GetTextureSwapChainBufferGL(session, eyes[eye].ColorTextureChain, curIndex, &colorTextureId);
+    }
+    {
+      int curIndex;
+      ovr_GetTextureSwapChainCurrentIndex(session, eyes[eye].DepthTextureChain, &curIndex);
+      ovr_GetTextureSwapChainBufferGL(session, eyes[eye].DepthTextureChain, curIndex, &depthTextureId);
+    }
+
+    glBindFramebuffer(GL_FRAMEBUFFER, fboId);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, colorTextureId, 0);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, depthTextureId, 0);
+
+    // glBlitNamedFramebuffer(
+    //   fboIdSource,
+    //   fboId,
+    //   eye == 0 ? 0 : canvasWidth / 2, 0,
+    //   eye == 0 ? canvasWidth / 2 : canvasWidth, canvasHeight,
+    //   0, 0, eyes[eye].textureSize.w, eyes[eye].textureSize.h,
+    //   GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT,
+    //   GL_LINEAR);
+
+    glViewport(0, 0, eyes[eye].textureSize.w, eyes[eye].textureSize.h);
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+    glEnable(GL_FRAMEBUFFER_SRGB);
+
+    ovr_CommitTextureSwapChain(session, eyes[eye].ColorTextureChain);
+    ovr_CommitTextureSwapChain(session, eyes[eye].DepthTextureChain);
+
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, 0, 0);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, 0, 0);
+
+    if (gl->HasTextureBinding(gl->activeTexture, GL_TEXTURE_2D)) {
```

Comment:
We are binding the framebuffer above, likely overwriting the user's current binding. Whenever we do something like this, we need to restore the user's binding back. That is what e.g. `gl->HasFramebufferBinding` is for.